src/permission-set.js

```diff
@@ -16,523 +16,809 @@
-    // An authorization for this agent and resource combination already exists
-    // Merge the incoming access modes with its existing ones
-    this.authorizations[hashFragment].mergeWith(auth)
-  } else {
-    this.authorizations[hashFragment] = auth
+  /**
+   * Adds a given authorization to the "lookup by group id" index.
+   * Enables lookups via `findAuthByAgent()`.
+   * @method addToGroupIndex
+   * @private
+   * @param webId {String} Group's Web ID
+   * @param accessType {String} Either `accessTo` or `default`
+   * @param resourceUrl {String}
+   * @param authorization {Authorization}
+   */
+  addToGroupIndex (webId, accessType, resourceUrl, authorization) {
+    let groups = this.index.groups
+    if (!groups[webId]) {
+      groups[webId] = {}
+    }
+    if (!groups[webId][accessType]) {
+      groups[webId][accessType] = {}
+    }
+    if (!groups[webId][accessType][resourceUrl]) {
+      groups[webId][accessType][resourceUrl] = authorization
+    } else {
+      groups[webId][accessType][resourceUrl].mergeWith(authorization)
+    }
   }
-  return this
-}
-PermissionSet.prototype.addAuthorization = addAuthorization
 
-/**
- * Adds an agentClass/group permission for the given access mode and agent id.
- * @method addGroupPermission
- * @param webId {String}
- * @param accessMode {String|Array<String>}
- * @return {PermissionSet} Returns self (chainable)
- */
-function addGroupPermission (webId, accessMode) {
-  var auth = new Authorization(this.resourceUrl, this.isAuthInherited())
-  auth.setGroup(webId)
-  auth.addMode(accessMode)
-  this.addAuthorization(auth)
-  return this
-}
-PermissionSet.prototype.addGroupPermission = addGroupPermission
+  /**
+   * Adds a given authorization to the "lookup by group id" index.
+   * Alias for `addToGroupIndex()`.
+   * Enables lookups via `findAuthByAgent()`.
+   * @method addToPublicIndex
+   * @private
+   * @param resourceUrl {String}
+   * @param accessType {String} Either `accessTo` or `default`
+   * @param authorization {Authorization}
+   */
+  addToPublicIndex (resourceUrl, accessType, auth) {
+    this.addToGroupIndex(acl.EVERYONE, accessType, resourceUrl, auth)
+  }
 
-/**
- * Adds a permission for the given access mode and agent id.
- * @method addPermission
- * @param webId {String} URL of an agent for which this permission applies
- * @param accessMode {String|Array<String>} One or more access modes
- * @param [origin] {String|Array<String>} One or more allowed origins (optional)
- * @return {PermissionSet} Returns self (chainable)
- */
-function addPermission (webId, accessMode, origin) {
-  if (!webId) {
-    throw new Error('addPermission() requires a valid webId')
+  /**
+   * Returns a list of all the Authorizations that belong to this permission set.
+   * Mostly for internal use.
+   * @method allAuthorizations
+   * @return {Array<Authorization>}
+   */
+  allAuthorizations () {
+    var authList = []
+    var auth
+    var self = this
```

Comment:
Some nitpicks: You don't need `self` (w/ es6 arrow functions). You could also make this a bit shorter by using `map` rather than `forEach`/imperatively mutating an array.